File: packages/fastapi/fastbackend_fastapi/validation_engine.py
"""Dynamic Pydantic model creation from IR."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field, create_model


TYPE_MAP: dict[str, Any] = {
    "string": str,
    "integer": int,
    "float": float,
    "boolean": bool,
    "date": date,
    "datetime": datetime,
    "json": dict,
    "uuid": UUID,
    "email": EmailStr,
}


class ValidationEngine:
    """Creates Pydantic models dynamically from IR entities."""

    def __init__(self) -> None:
        self._models: dict[str, type[BaseModel]] = {}

    def create_models(self, entity: dict[str, Any]) -> dict[str, type[BaseModel]]:
        name = entity["name"]
        full = self._create_model(entity, f"{name}", mode="full")
        create = self._create_model(entity, f"{name}Create", mode="create")
        update = self._create_model(entity, f"{name}Update", mode="update")

        self._models[name] = full
        self._models[f"{name}Create"] = create
        self._models[f"{name}Update"] = update

        return {"full": full, "create": create, "update": update}

    def get_model(self, name: str) -> type[BaseModel] | None:
        return self._models.get(name)

    def _create_model(
        self, entity: dict[str, Any], model_name: str, mode: str
    ) -> type[BaseModel]:
        fields: dict[str, Any] = {}
        pk = entity.get("primaryKey", ["id"])

        for field_def in entity["fields"]:
            field_name = field_def["name"]

            if mode == "create" and field_name in pk and field_name == "id":
                continue

            py_type = self._map_type(field_def)
            field_kwargs: dict[str, Any] = {}

            if mode == "update":
                py_type = Optional[py_type]
                field_kwargs["default"] = None
            elif field_def.get("nullable", True) and mode != "create":
                py_type = Optional[py_type]
                field_kwargs["default"] = None
            elif not field_def.get("nullable", True) and mode == "create":
                is_required = any(
                    v.get("type") == "required" for v in field_def.get("validation", [])
                )
                if not is_required and field_def.get("defaultValue") is None:
                    field_kwargs["default"] = ...

            for rule in field_def.get("validation", []):
                if rule["type"] == "minLength" and "value" in rule:
                    field_kwargs["min_length"] = rule["value"]
                elif rule["type"] == "maxLength" and "value" in rule:
                    field_kwargs["max_length"] = rule["value"]
                elif rule["type"] == "min" and "value" in rule:
                    field_kwargs["ge"] = rule["value"]
                elif rule["type"] == "max" and "value" in rule:
                    field_kwargs["le"] = rule["value"]
                elif rule["type"] == "pattern" and "value" in rule:
                    field_kwargs["pattern"] = rule["value"]

            if field_kwargs:
                fields[field_name] = (py_type, Field(**field_kwargs))
            else:
                fields[field_name] = (py_type, ...)

        return create_model(model_name, **fields)

    def _map_type(self, field_def: dict[str, Any]) -> Any:
        field_type = field_def.get("type", {})
        base = field_type.get("base", "string")
        fmt = field_type.get("format")

        if fmt == "email":
            return EmailStr
        if fmt == "uuid":
            return UUID
        if base == "array" and field_type.get("arrayOf"):
            inner = self._map_type({"type": field_type["arrayOf"]})
            return list[inner]

        return TYPE_MAP.get(base, str)

File: packages/express/src/engines/validation-engine.ts
import { z } from 'zod';
import type { IRField, IREntity } from '../utils/ir-loader.js';

export class ValidationEngine {
  private readonly schemas = new Map<string, z.ZodTypeAny>();

  createSchemas(entity: IREntity): void {
    const full = this.buildSchema(entity, 'full');
    const create = this.buildSchema(entity, 'create');
    const update = this.buildSchema(entity, 'update');

    this.schemas.set(entity.name, full);
    this.schemas.set(`${entity.name}Create`, create);
    this.schemas.set(`${entity.name}Update`, update);
  }

  getSchema(name: string): z.ZodTypeAny | undefined {
    return this.schemas.get(name);
  }

  validatePayload(name: string, body: unknown, partial = false): Record<string, unknown> {
    const schema = this.getSchema(name);
    if (!schema) {
      if (typeof body !== 'object' || body === null || Array.isArray(body)) {
        throw new ValidationError('Request body must be a JSON object');
      }
      return body as Record<string, unknown>;
    }

    const validator = partial && schema instanceof z.ZodObject ? schema.partial() : schema;
    const parsed = validator.safeParse(body);
    if (!parsed.success) {
      throw new ValidationError('Validation failed', parsed.error.flatten());
    }

    return parsed.data as Record<string, unknown>;
  }

  private buildSchema(entity: IREntity, mode: 'full' | 'create' | 'update'): z.ZodObject<Record<string, z.ZodTypeAny>> {
    const pk = entity.primaryKey ?? ['id'];
    const shape: Record<string, z.ZodTypeAny> = {};

    for (const field of entity.fields) {
      if (mode === 'create' && pk.includes(field.name) && field.name === 'id') {
        continue;
      }

      let fieldSchema = this.mapField(field);

      if (mode === 'update') {
        fieldSchema = fieldSchema.optional();
      } else if (field.nullable && mode !== 'create') {
        fieldSchema = fieldSchema.nullable().optional();
      } else if (!field.nullable && mode === 'create') {
        const required = field.validation.some((rule) => rule.type === 'required');
        if (!required && field.defaultValue === undefined) {
          fieldSchema = fieldSchema.optional();
        }
      }

      for (const rule of field.validation) {
        fieldSchema = this.applyRule(fieldSchema, rule, field);
      }

      shape[field.name] = fieldSchema;
    }

    return z.object(shape);
  }

  private mapField(field: IRField): z.ZodTypeAny {
    const base = field.type.base;
    const format = field.type.format;

    switch (base) {
      case 'integer':
        return z.number().int();
      case 'float':
        return z.number();
      case 'boolean':
        return z.boolean();
      case 'date':
      case 'datetime':
        return z.string();
      case 'json':
        return z.record(z.unknown());
      case 'array':
        return z.array(this.mapField({ ...field, type: field.type.arrayOf ?? { base: 'string' } }));
      case 'string':
        if (format === 'email') {
          return z.string().email();
        }
        if (format === 'uuid') {
          return z.string().uuid();
        }
        return z.string();
      default:
        return z.string();
    }
  }

  private applyRule(schema: z.ZodTypeAny, rule: { type: string; value?: unknown }, field: IRField): z.ZodTypeAny {
    switch (rule.type) {
      case 'minLength':
        return schema instanceof z.ZodString ? schema.min(Number(rule.value)) : schema;
      case 'maxLength':
        return schema instanceof z.ZodString ? schema.max(Number(rule.value)) : schema;
      case 'min':
        return schema instanceof z.ZodNumber ? schema.min(Number(rule.value)) : schema;
      case 'max':
        return schema instanceof z.ZodNumber ? schema.max(Number(rule.value)) : schema;
      case 'pattern':
        return schema instanceof z.ZodString ? schema.regex(new RegExp(String(rule.value))) : schema;
      case 'required':
        return field.nullable ? schema : schema;
      default:
        return schema;
    }
  }
}

export class ValidationError extends Error {
  constructor(
    message: string,
    readonly details?: unknown,
  ) {
    super(message);
    this.name = 'ValidationError';
  }
}
